### ISSUE TITLE
Connection dying after a few minutes - Semaphore Timeout Error 

### ISSUE THREAD

**Issue by JulianDeal** [aaugustin/websockets#924]
I got this error log after having a websocket client open for several minutes.

I'm receiving normal pings, and then some '\x92\xd8m\x9d' and then websockets crashes.

```
[1616527748.531067][websockets.protocol][DEBUG] client - event = data_received(<6 bytes>)
[1616527748.531067][websockets.protocol][DEBUG] client < Frame(fin=True, opcode=1, data=b'ping', rsv1=False, rsv2=False, rsv3=False)
[1616527748.595066][websockets.protocol][DEBUG] client - event = data_received(<6 bytes>)
[1616527748.5960674][websockets.protocol][DEBUG] client < Frame(fin=True, opcode=1, data=b'ping', rsv1=False, rsv2=False, rsv3=False)
[1616527750.7327232][websockets.protocol][DEBUG] client > Frame(fin=True, opcode=9, data=b'\x92\xd8m\x9d', rsv1=False, rsv2=False, rsv3=False)
[1616527769.9952965][websockets.protocol][DEBUG] client - event = connection_lost([WinError 121] Das Zeitlimit für die Semaphore wurde erreicht)
[1616527769.9962957][websockets.protocol][DEBUG] client - state = CLOSED
[1616527769.9962957][websockets.protocol][DEBUG] client x code = 1006, reason = [no reason]
[1616527769.9972994][websockets.protocol][DEBUG] client - aborted pending ping: 92d86d9d
[1616527770.003296][websockets.protocol][ERROR] Error in data transferTraceback (most recent call last):
  File "C:\…\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\websockets\protocol.py", line 827, in transfer_data
    message = await self.read_message()
  File "C:\…\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\websockets\protocol.py", line 895, in read_message
    frame = await self.read_data_frame(max_size=self.max_size)
  File "C:\…\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\websockets\protocol.py", line 971, in read_data_frame
    frame = await self.read_frame(max_size)
  File "C:\…\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\websockets\protocol.py", line 1047, in read_frame
    frame = await Frame.read(
  File "C:\…\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\websockets\framing.py", line 105, in read
    data = await reader(2)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\streams.py", line 723, in readexactly
    await self._wait_for_data('readexactly')
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\streams.py", line 517, in _wait_for_data
    await self._waiter
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\futures.py", line 284, in __await__
    yield self  # This tells Task to wait for completion.
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\tasks.py", line 328, in __wakeup
    future.result()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\futures.py", line 201, in result
    raise self._exception
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line 280, in _loop_reading
    data = fut.result()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\windows_events.py", line 812, in _poll
    value = callback(transferred, key, ov)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.752.0_x64__qbz5n2kfra8p0\lib\asyncio\windows_events.py", line 457, in finish_recv
    return ov.getresult()
OSError: [WinError 121] Das Zeitlimit für die Semaphore wurde erreicht
[1616527770.0102963][websockets.protocol][DEBUG] client ! failing CLOSED WebSocket connection with code 1011
[1616527770.0102963][websockets.protocol][DEBUG] client x closing TCP 
connection
```

Is there a way to handle that exception or re-establish the connection in the meanwhile?

**Comment by aaugustin** [comment]
This looks like a duplicate of #810.

**Comment by JulianDeal** [comment]
I'm quite new to all of this and a bit confused. So please bear with me.

Sorry, I blindly assumed the change in #810 was committed to the official package because half a year passed . Nevertheless, I don't see how the change prevents the connection from failing. `self.fail_connection(1006)` is still called.

I don't get this error anymore when using WSL2 (Ubuntu 20.04), at least I could run my script the whole day yesterday. But something similar happened today, again. (But I couldn't fetch the log due to a power failure while trying to copy and paste the log). 

Looking at the original issue again: `\x92\xd8m\x9d`. I still don't know what `\x8dm` should represent. All the received messages are HTML with UTF8-encoded as JSON and beginning with `<tr` or a simple `ping` (which all seems a bit strange to me but I'm not the author of the server).

Aside all of this. I'm asking for some guidance here: 

- What is the proper way to handle such errors and reconnect to the server if the case? 
- And are there known ways to reproduce similar issues?

**Comment by aaugustin** [comment]
`\x92\xd8m\x9d` is just 4 random bytes put in the payload of a ping. They have no special meaning. It's a control frame. You don't need to care about this, but you can disable it with `ping_timeout=None` if you think it's part of the problem.

**Comment by aaugustin** [comment]
I made a release yesterday that includes the fix for #810. At this point I'm not seeing any remaining action so I'm going to close the issue.
